client/asset/dcr/coin_selection.go

```diff
@@ -71,69 +72,67 @@ func sumUTXOs(set []*compositeUTXO) (tot uint64) {
 	return tot
 }
 
-// In the following utxo selection functions, the compositeUTXO slice MUST be
-// sorted in ascending order (smallest first, largest last).
-
-// subsetLargeBias begins by summing from the largest UTXO down until the sum is
-// just below the requested amount. Then it picks the (one) final UTXO that
-// reaches the amount with the least excess. Each utxo in the input must be
-// smaller than amt - use via leastOverFund only!
-func subsetLargeBias(amt uint64, utxos []*compositeUTXO) []*compositeUTXO {
-	// Add from largest down until sum is *just under*. Resulting sum will be
-	// less, and index will be the next (smaller) element that would hit amt.
-	var sum uint64
-	var i int
-	for i = len(utxos) - 1; i >= 0; i-- {
-		this := toAtoms(utxos[i].rpc.Amount) // must not be >= amt
-		if sum+this >= amt {
-			break
+// subsetWithLeastSumGreaterThan attempts to select the subset of UTXOs with
+// the smallest total value greater than amt. It does this by making
+// 1000 random selections and returning the best one. Each selection
+// involves two passes over the UTXOs. The first pass randomly selects
+// each UTXO with 50% probability. Then, the second pass selects any
+// unused UTXOs until the total value is greater than or equal to amt.
+func subsetWithLeastSumGreaterThan(amt uint64, utxos []*compositeUTXO) []*compositeUTXO {
+	best := uint64(1 << 62)
+	var bestIncluded []bool
+	bestNumIncluded := 0
+
+	rnd := rand.New(rand.NewSource(rand.Int63()))
+	included := make([]bool, len(utxos))
+	const iterations = 1000
+	for nRep := 0; nRep < iterations; nRep++ {
+		var nTotal uint64
+		var numIncluded int
+	passes:
+		for nPass := 0; nPass < 2; nPass++ {
+			for i := 0; i < len(utxos); i++ {
+				var use bool
+				if nPass == 0 {
+					use = rnd.Int63()&1 == 1
+				} else {
+					use = !included[i]
+				}
+				if use {
```

Comment:
It still bugs me that there's no ternary operator in Go.  The arguments against it are so lame.
We could do this, but it's less clear and concise: `if (nPass == 0 && rnd.Int63()&1 == 1) || (nPass > 0 && !included[i]) {`.
We'll never get a ternary and probably not a context.Merge either.